pkg/controller/velero/controller.go

```diff
@@ -0,0 +1,172 @@
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to VolumeSnapshotLocation
+	err = c.Watch(&source.Kind{Type: &velerov1.VolumeSnapshotLocation{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CredentialsRequest
+	err = c.Watch(&source.Kind{Type: &minterv1.CredentialsRequest{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Deployments
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileVelero implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileVelero{}
+
+// ReconcileVelero reconciles a Velero object
+type ReconcileVelero struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Velero object and makes changes based on the state read
+// and what is in the Velero.Spec
+func (r *ReconcileVelero) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Velero Installation")
+	var err error
+
+	// Fetch the Velero instance
+	instance := &veleroCR.Velero{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Grab platform status to determine where OpenShift is installed
+	platformStatus, err := platform.GetPlatformStatus(r.client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Verify that we have received an AWS region from the platform
+	if platformStatus.AWS == nil || len(platformStatus.AWS.Region) < 1 {
+		return reconcile.Result{}, fmt.Errorf("unable to determine AWS region")
+	}
+
+	// Create an S3 client based on the region we received
+	s3Client, err := s3.NewS3Client(r.client, platformStatus.AWS.Region)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// If any of the following are true, reconcile the S3 bucket:
+	// - Name is empty
+	// - Provisioned is false
+	// - The LastSyncTimestamp is unset
+	// - It's been longer than 1 hour since last sync
+	if instance.Status.S3Bucket.Name == "" ||
```

Comment:
Thoughts on a method for instances to consolidate all of this logic? So it might look like `instance.PerformReconcile(...)`?

Reply:
I like that idea!

Reply:
Done